# EDA on Waze User Churn Data Phase-I

My Waze data analytics team is currently in the initial stages of our user churn project. After submitting a project proposal to my supervisor, May Santner, I received approval along with access to Waze’s user data. To gain accurate insights, our first step is to inspect and prepare this data for the upcoming exploratory data analysis (EDA) phase.


**Purpose:**  
To investigate and understand the data provided.

**Goal:**  
To construct a DataFrame in Python, perform an initial inspection of the dataset, and share key findings with team members.

The project is structured in three main tasks:

**Task 1: Understanding the Context**  
- Organize and interpret the information provided to fully grasp the project scope.

**Task 2: Understanding the Data**  
- Build a pandas DataFrame for data exploration, EDA, and statistical analysis.  
- Compile summary insights to inform subsequent steps.

**Task 3: Understanding the Variables**  
- Leverage insights from data summaries to conduct a deeper investigation into individual variables. 

This structured approach will lay a strong foundation for effective analysis and enable us to proactively address user churn.

# **Identify data types and compile summary information**


<img src="assets/Pace.png" width="100" height="100" align=left>

# **PACE stages**

Throughout these project notebooks, references to the problem-solving framework, PACE, are included. Each notebook component is labeled according to its corresponding PACE stage: Plan, Analyze, Construct, and Execute.

<img src="assets/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

### **Task 1. Understanding the situation**

Exploring the dataset, reviewing the data dictionary, and examining user churn data fields to understand the data and the impact of each field, as well as reviewing the fact sheet, can provide essential background information. 

The main goal is to load the dataset into a Python dataframe, inspect it, assess its quality, and provide May, the Data Analysis Manager, with initial observations. 

Next steps will involve gaining a deeper understanding of the data and checking for any anomalies.


<img src="assets/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

### **Task 2a. Imports and data loading**

Starting by importing the packages that will be needed to load and explore the dataset.

In [81]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

Then, loading the dataset into a dataframe. Creating a dataframe will help conducting data manipulation, exploratory data analysis (EDA), and statistical activities.

In [83]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Task 2b. Summary information**

Inspecting summary information about the dataframe .


In [85]:
#Display first 10 rows.
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [86]:
# Display a concise summary of the DataFrame, including the number of non-null values,
# data types for each column, and overall memory usage. 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


- There are "3 of float64" , "8 of int64" , "2 of object" and there are "13" column and "14999" rows
- There are 700 missing values in the "label" column in the dataset.

### **Task 2c. Null values and summary statistics**

Comparing the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.


In [89]:
# Isolate rows with null values
null_lable_df=df[df["label"].isnull()]
# Display summary stats of rows with null values
null_lable_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [90]:
# Isolate rows without null values
not_null_lable_df=df[~df["label"].isnull()]

# Display summary stats of rows without null values
not_null_lable_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

The mean values are nearly equal, but there are differences in the minimum values of the number of days since the user signed up, the kilometers driven, and the minutes driven. The `label` column with missing values has higher minimum values in these fields compared to the non-null `label`. However, for maximum values, the `label` column with missing values is lower in these fields than the non-null `label`.

There are anomalies in the number of days since the user signed up, the kilometers driven, and the minutes driven in both the non-null `label` and missing values `label`. The number of sessions is similar across each type of `label`, despite the non-null `label` containing nearly twice the number of rows as the missing values `label`. This might indicate that the `label` column with missing values contains data for earlier users.

### **Task 2d. Null values - device counts**

Checking the two populations with respect to the `device` variable.

In [93]:
# Get count of null values by device
null_lable_df["device"].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

`iPhone`:`447` and `Android`:`253`, In the `700` rows with label value is missing

Now, of the rows with null values, calculate the percentage with each device&mdash;Android and iPhone. You can do this directly with the [`value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) function.

In [96]:
# Calculate % of iPhone nulls and Android nulls
null_lable_df['device'].value_counts(normalize=True)

device
iPhone     0.638571
Android    0.361429
Name: proportion, dtype: float64

In [97]:
# Calculate % of iPhone users and Android users in full dataset
not_null_lable_df["device"].value_counts(normalize=True)

device
iPhone     0.64515
Android    0.35485
Name: proportion, dtype: float64

- The percentage of missing values by each device is consistent with their representation in the data overall.

- There is nothing to suggest a non-random cause of the missing data.

Examining the counts and percentages of users who churned vs. those who were retained.

In [100]:
# Calculate counts of churned vs. retained
print("Churned: "+str(not_null_lable_df['label'].value_counts(normalize=True)['churned']*100)+" %")
print("Retained: "+str(not_null_lable_df['label'].value_counts(normalize=True)['retained']*100)+" %")

Churned: 17.73550597943912 %
Retained: 82.26449402056087 %


- This dataset contains `82%` `retained` users and `18%` `churned` users.

Comparing the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that I don't want outliers to unduly affect the portrayal of a typical user. For example, that the maximum value in the `driven_km_drives` column is 21,183 km. 

In [112]:
# Select only numeric columns for median calculation
numeric_cols = df.select_dtypes(include=['number']).columns

# Calculate median values of numeric columns for churned and retained users
median_values = df.groupby(["label"])[numeric_cols].median()
median_values

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

This offers an interesting snapshot of the two groups, `churned` vs. `retained`:

Users who churned averaged ~`3` more drives in the last month than `retained` users, but `retained` users used the app on over twice as many days as `churned` users in the same time period.

The median `churned` user drove ~`200` more kilometers and `2.5` more `hours` during the last month than the median `retained` user.

It seems that `churned` users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile.

Calculating the median kilometers per drive in the last month for both `retained` and `churned` users.

In [114]:
# Divide the median distance by median number of drives
median_values["driven_km_drives"]/median_values["drives"]

label
churned     73.053113
retained    73.716694
dtype: float64

The median user from both groups drove ~73 km/drive. 

In [116]:
# Divide the median distance by median number of driving days
median_values["driven_km_drives"]/median_values["driving_days"]

label
churned     608.775944
retained    247.477472
dtype: float64

Calculating the median number of drives per driving day for each group.

In [118]:
# Divide the median number of drives by median number of driving days
median_values["drives"]/median_values["driving_days"]

label
churned     8.333333
retained    3.357143
dtype: float64

`Churned` users drive significantly more than `retained` users, with a `median` of `608 km` per day—about `250%` of the distance `retained` users drive per active day. They also show a disproportionate number of drives per day.

These data likely represent dedicated drivers, possibly including many long-haul truckers, and may not reflect typical drivers. It might be beneficial for Waze to collect more insights on these high-mileage users, as their needs may differ from those of casual drivers and could explain why the app may not fully address their expectations.

Examining whether there is an imbalance in how many users churned by device type.

Begining by getting the overall counts of each device type for each group, churned and retained.

In [120]:
# For each label, calculate the number of Android users and iPhone users
df.groupby(["label","device"]).count()[["ID"]]

ID
label    device       
churned  Android   891
         iPhone   1645
retained Android  4183
         iPhone   7580

Calculating what percent was Android and what percent was iPhone, within each group, churned and retained.

In [122]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby(["label"])["device"].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

<img src="assets/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.



<img src="assets/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

### **Task 3. Conclusion**

- [Executive Summary](https://docs.google.com/presentation/d/12WrMK0GGLcDfY8RrOpu5qq8GxgBxJsmBDJT_bjpUJ7o/edit?usp=sharing)

### Key Insights for Stakeholders

1. **User Retention Overview:**
   - The dataset reveals that 82% of users are retained, while 18% have churned. This distribution highlights the importance of understanding the characteristics and behaviors of both groups to enhance user retention strategies.

2. **Missing Values:**
   - The "label" column contains 700 missing values, with no evidence suggesting that these omissions are non-random. This could indicate a potential bias in the data, particularly related to early users. Further investigation into these missing values is essential.

3. **Driving Behavior Patterns:**
   - Churned users averaged about three more drives in the last month compared to retained users, yet retained users used the app on more than twice as many days. This suggests that while churned users may drive more frequently in a shorter time frame, their overall app engagement is lower.

4. **Comparative Analysis of Churned vs. Retained Users:**
   - The median churned user drove approximately 698 kilometers per day—240% more than retained users—and spent an additional 2.5 hours driving over the last month. This indicates that churned users not only drive farther but also longer during each session.

5. **User Demographics:**
   - The dataset comprises 12 unique variables, including objects, floats, and integers. Android users constitute about 36% of the sample, while iPhone users account for 64%. This information can help tailor app features and marketing efforts to specific user demographics.

6. **Churn Rate Analysis:**
   - The churn rate between iPhone and Android users is comparable, differing by only one percentage point, which suggests that device type does not significantly impact user retention.

7. **Implications of Driving Behavior:**
   - Regardless of user churn status, the data indicates that users drive extensively. This raises questions about whether this sample represents typical drivers, as the high mileage may correlate with specific user profiles, such as long-haul truckers. Continued exploration of this user profile could yield insights for enhancing app functionality and user experience.

### Summary
This analysis provides crucial insights into user behavior, retention rates, and data characteristics, emphasizing the need for further exploration of the driving patterns and preferences of both churned and retained users. By understanding these dynamics, Waze can develop targeted strategies to improve user retention and better meet the needs of its diverse user base.

